electrum/lnworker.py

```diff
@@ -1836,12 +1837,24 @@ def get_bolt11_invoice(
             + routing_hints
             + trampoline_hints,
             date=timestamp,
-            payment_secret=derive_payment_secret_from_payment_preimage(payment_preimage))
+            payment_secret=payment_secret)
         invoice = lnencode(lnaddr, self.node_keypair.privkey)
         pair = lnaddr, invoice
         self._bolt11_cache[payment_hash] = pair
         return pair
 
+    def get_payment_secret(self, payment_hash, create=False):
+        preimage = self.get_preimage(payment_hash)
+        derived_secret = derive_payment_secret_from_payment_preimage(preimage) if preimage else None
+        if create and not preimage:
+            self._payment_secrets[payment_hash] = os.urandom(32)
```

Comment:
I think it would be possible to not have to store payment secrets but derive them from other secret data, e.g.
`payment_secret = H(H(H("payment_secret") || H(node_privkey)) || payment_hash)`
Perhaps we could even just switch over completely to the new scheme, even for invoices where we have the preimage.
If we don't want to break the usecase where a user has generated and given out a payment request and then upgraded their client and then is receiving payment, then in lnpeer.maybe_fulfill_htlc we could maybe derive both payment_secrets and accept either.